### PR TITLE
Provide UI support to configure SP level id token encryption algorithms

### PR DESCRIPTION
**Pull request by vihanga-liyanage**
### Proposed changes in this pull request

- ID Token encryption algorithms can be configured per each SP using the UI as shown below.
image

- Drop down menu's and default algorithms are populated by reading the identity.xml file. These changes are added in the PR [wso2/carbon-identity-framework#1416](https://github.com/wso2/carbon-identity-framework/pull/1416)

Fix [wso2/product-is#2387](https://github.com/wso2/product-is/issues/2387)

### When should this PR be merged

No dependencies, hence ASAP

#### General

- [ ] **Is this PR explained thoroughly?** All code changes must be accounted for in the PR description.
- [ ] **Is the PR labeled correctly?**

#### Functionality

- [ ] **Are all requirements met?** Compare implemented functionality with the requirements specification.
- [ ] **Does the UI work as expected?** There should be no Javascript errors in the console; all resources should load. There should be no unexpected errors. Deliberately try to break the feature to find out if there are corner cases that are not handled.

#### Code

- [ ] **Do you fully understand the introduced changes to the code?** If not ask for clarification, it might uncover ways to solve a problem in a more elegant and efficient way.
- [ ] **Does the PR introduce any inefficient database requests?** Use the debug server to check for duplicate requests.
- [ ] **Are all necessary strings marked for translation?** All strings that are exposed to users via the UI must be [marked for translation](https://docs.djangoproject.com/en/1.10/topics/i18n/translation/).

#### Tests

- [ ] **Are there sufficient test cases?** Ensure that all components are tested individually; models, forms, and serializers should be tested in isolation even if a test for a view covers these components.
- [ ] **If this is a bug fix, are tests for the issue in place?**  There must be a test case for the bug to ensure the issue won’t regress. Make sure that the tests break without the new code to fix the issue.
- [ ] **If this is a new feature or a significant change to an existing feature?** has the manual testing spreadsheet been updated with instructions for manual testing?

#### Security

- [ ] **Confirm this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets.**
- [ ] **Are all UI and API inputs run through forms or serializers?**
- [ ] **Are all external inputs validated and sanitized appropriately?**
- [ ] **Does all branching logic have a default case?**
- [ ] **Does this solution handle outliers and edge cases gracefully?**
- [ ] **Are all external communications secured and restricted to SSL?**

#### Documentation

- [ ] **Are changes to the UI documented in the platform docs?** If this PR introduces new platform site functionality or changes existing ones, the changes should be documented.
- [ ] **Are changes to the API documented in the API docs?** If this PR introduces new API functionality or changes existing ones, the changes must be documented.
- [ ] **Are reusable components documented?** If this PR introduces components that are relevant to other developers (for instance a mixin for a view or a generic form) they should be documented in the Wiki.

### CODE REVIEW

File: components/org.wso2.carbon.identity.oauth.stub/src/main/resources/OAuthAdminService.wsdl [hunk 1]
```diff
@@ -11,27 +11,27 @@
                     </xs:sequence>
                 </xs:complexType>
             </xs:element>
-            <xs:element name="getAllOAuthApplicationData">
+            <xs:element name="registerOAuthConsumer">
```
Comment (ruwanta) [inline components/org.wso2.carbon.identity.oauth.stub/src/main/resources/OAuthAdminService.wsdl]: Can we revert these unnecessary changes and just retain "encryptionAlgorithm" "encryptionMethod" element additions only. It make bit harder to track the changes.

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/config/OAuthServerConfiguration.java [hunk 2]
```diff
@@ -2096,6 +2123,48 @@ private void parseOpenIDConnectConfig(OMElement oauthConfigElem) {
...
+        if (algorithms == null) {
+            return;
+        }
+
+        Iterator iterator = algorithms.getChildrenWithLocalName(ConfigElements.SUPPORTED_IDTOKEN_ENCRYPTION_ALGORITHM);
+        if (iterator != null) {
+            for (; iterator.hasNext(); ) {
+                OMElement algorithm = (OMElement) iterator.next();
+                if (algorithm != null) {
+                    log.info(algorithm.getText());
```
Comment (GayanM) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/config/OAuthServerConfiguration.java]: is it make sense to just log algorithm.getText()

Comment (vihanga-liyanage) [reply components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/config/OAuthServerConfiguration.java]: That was mistakenly left there. I'll remove it ASAP.

File: components/org.wso2.carbon.identity.oauth.ui/src/main/resources/web/oauth/add-finish-ajaxprocessor.jsp [hunk 3]
```diff
@@ -71,7 +71,9 @@
     // OIDC related properties
     boolean isRequestObjectSignatureValidated = Boolean.parseBoolean(request.getParameter("validateRequestObjectSignature"));
     boolean isIdTokenEncrypted = Boolean.parseBoolean(request.getParameter("encryptIdToken"));
-    
+    String encryptionAlgorithm = request.getParameter("encryptionAlgorithm");
```
Comment (nilasini) [inline components/org.wso2.carbon.identity.oauth.ui/src/main/resources/web/oauth/add-finish-ajaxprocessor.jsp]: Can't we use the constants defined in OAuthConstants.java for encryptionAlgorithm & encryptionMethod

File: components/org.wso2.carbon.identity.oauth.ui/src/main/resources/web/oauth/edit-finish-ajaxprocessor.jsp [hunk 4]
```diff
@@ -73,6 +73,8 @@
     // OIDC related properties
     boolean isRequestObjectSignatureValidated = Boolean.parseBoolean(request.getParameter("validateRequestObjectSignature"));
     boolean isIdTokenEncrypted = Boolean.parseBoolean(request.getParameter("encryptIdToken"));
+    String encryptionAlgorithm = request.getParameter("encryptionAlgorithm");
+    String encryptionMethod = request.getParameter("encryptionMethod");
```
Comment (nilasini) [inline components/org.wso2.carbon.identity.oauth.ui/src/main/resources/web/oauth/edit-finish-ajaxprocessor.jsp]: Can't we use already defined constants?

File: components/org.wso2.carbon.identity.oauth.ui/src/main/resources/web/oauth/edit.jsp [hunk 5]
```diff
@@ -391,6 +406,10 @@
                     //on load adjust the form based on the current settings
                     adjustForm();
                     $("form[name='editAppform']").change(adjustForm);
+
+                    // Set selected encryption algorithm and encryption method.
+                    $('select[name=encryptionAlgorithm]').val('<%=app.getEncryptionAlgorithm()%>');
+                    $('select[name=encryptionMethod]').val('<%=app.getEncryptionMethod()%>');
                 })
```
Comment (malithie) [inline components/org.wso2.carbon.identity.oauth.ui/src/main/resources/web/oauth/edit.jsp]: Entries for 'app' reference is being set from request parameters. Thus, these values needs to be properly encoded. Maybe, jquery handles this. Yet, have a check and confirm.

File: components/org.wso2.carbon.identity.oauth.ui/src/main/resources/web/oauth/edit.jsp [hunk 6]
```diff
@@ -67,6 +68,11 @@
     List<String> allowedScopeValidators = new ArrayList<String>();
     List<String> scopeValidators = new ArrayList<String>();
 
+    List<String> supportedEncryptionAlgorithms =
+            OAuthServerConfiguration.getInstance().getSupportedIdTokenEncryptionAlgorithm();
+    List<String> supportedEncryptionMethods =
+            OAuthServerConfiguration.getInstance().getSupportedIdTokenEncryptionMethods();
+
```
Comment (malithie) [inline components/org.wso2.carbon.identity.oauth.ui/src/main/resources/web/oauth/edit.jsp]: Here backend APIs are directly used adding dependencies for the backend component. Shouldn't this be invoked over a service API. Could you please check the common practice being used in other UI bundles

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/config/OAuthServerConfiguration.java [hunk 7]
```diff
@@ -2096,6 +2123,46 @@ private void parseOpenIDConnectConfig(OMElement oauthConfigElem) {
         }
     }
 
+    /**
+     * Parse supported encryption algorithms set and add them to supportedIdTokenEncryptionAlgorithms.
+     * @param algorithms   OMElement of supported algorithms.
+     */
+    private void parseSupportedIdTokenEncryptionAlgorithms(OMElement algorithms) {
```
Comment (malithie) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/config/OAuthServerConfiguration.java]: Usually there's a new line added after the method signature for clarity. If you have applied appropriate formatter it should be formatted so by default.

File: components/org.wso2.carbon.identity.oauth.ui/src/main/resources/web/oauth/add.jsp [hunk 8]
```diff
@@ -46,6 +47,13 @@
     String applicationSPName = request.getParameter("spName");
     session.setAttribute("application-sp-name", applicationSPName);
 
+    String encryptionAlgorithm = OAuthServerConfiguration.getInstance().getIdTokenEncryptionAlgorithm();
```
Comment (denuwanthi) [inline components/org.wso2.carbon.identity.oauth.ui/src/main/resources/web/oauth/add.jsp]: Here in UI side we are directly calling server side classes (OAuthServerConfiguration).
In design wise I think iti is not right.
Is it possible to use UI client and call the server side classes?

Comment (vihanga-liyanage) [reply components/org.wso2.carbon.identity.oauth.ui/src/main/resources/web/oauth/add.jsp]: I have changed the approach. Please review. :)

File: components/org.wso2.carbon.identity.oauth.ui/src/main/resources/web/oauth/edit.jsp (hunk 6 above)
Comment (vihanga-liyanage) [reply components/org.wso2.carbon.identity.oauth.ui/src/main/resources/web/oauth/edit.jsp]: I have changed the approach. Please review. :)

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/config/OAuthServerConfiguration.java (hunk 7 above)
Comment (vihanga-liyanage) [reply components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/config/OAuthServerConfiguration.java]: Changed.

File: components/org.wso2.carbon.identity.oauth.ui/src/main/resources/web/oauth/edit-finish-ajaxprocessor.jsp (hunk 4 above)
Comment (vihanga-liyanage) [reply components/org.wso2.carbon.identity.oauth.ui/src/main/resources/web/oauth/edit-finish-ajaxprocessor.jsp]: These are actually the name tags defined in JSP files. I suppose we could import the constant class into the JSPs and use the constants but I don't see any actual use in that. Common practice is to use these as it is. WDYT?

File: components/org.wso2.carbon.identity.oauth.ui/src/main/resources/web/oauth/add-finish-ajaxprocessor.jsp (hunk 3 above)
Comment (vihanga-liyanage) [reply components/org.wso2.carbon.identity.oauth.ui/src/main/resources/web/oauth/add-finish-ajaxprocessor.jsp]: These are actually the name tags defined in JSP files. I suppose we could import the constant class into the JSPs and use the constants but I don't see any actual use in that. Common practice is to use these as it is. WDYT?

File: components/org.wso2.carbon.identity.oauth.ui/src/main/resources/web/oauth/edit.jsp (hunk 5 above)
Comment (vihanga-liyanage) [reply components/org.wso2.carbon.identity.oauth.ui/src/main/resources/web/oauth/edit.jsp]: Values encoded.

File: components/org.wso2.carbon.identity.oauth.ui/src/main/resources/web/oauth/edit.jsp [hunk 9]
```diff
@@ -391,6 +406,10 @@
                     //on load adjust the form based on the current settings
                     adjustForm();
                     $("form[name='editAppform']").change(adjustForm);
+
+                    // Set selected encryption algorithm and encryption method.
+                    $('select[name=encryptionAlgorithm]').val('<%=Encode.forHtmlAttribute(app.getEncryptionAlgorithm())%>');
+                    $('select[name=encryptionMethod]').val('<%=Encode.forHtmlAttribute(app.getEncryptionMethod())%>');
```
Comment (malithie) [inline components/org.wso2.carbon.identity.oauth.ui/src/main/resources/web/oauth/edit.jsp]: Did we check how jquery behaves. Also, this is a JS function. If this is to be encoded it should be for javascript.
Please revalidate

File: components/org.wso2.carbon.identity.oauth.ui/pom.xml [hunk 10]
```diff
@@ -83,6 +83,11 @@
             <groupId>org.wso2.eclipse.osgi</groupId>
             <artifactId>org.eclipse.osgi.services</artifactId>
         </dependency>
+        <dependency>
+            <groupId>org.wso2.carbon.identity.inbound.auth.oauth2</groupId>
+            <artifactId>org.wso2.carbon.identity.oauth</artifactId>
+            <scope>compile</scope>
+        </dependency>
```
Comment (malithie) [inline components/org.wso2.carbon.identity.oauth.ui/pom.xml]: Is this dependency needed any more ?

Comment (vihanga-liyanage) [reply components/org.wso2.carbon.identity.oauth.ui/pom.xml]: Removed.

File: components/org.wso2.carbon.identity.oauth.ui/src/main/resources/web/oauth/edit.jsp (hunk 9 above)
Comment (vihanga-liyanage) [reply components/org.wso2.carbon.identity.oauth.ui/src/main/resources/web/oauth/edit.jsp]: Changed for javascript.

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/config/OAuthServerConfiguration.java [hunk 11]
```diff
@@ -2046,18 +2058,39 @@ private void parseOpenIDConnectConfig(OMElement oauthConfigElem) {
                                 .getText().trim();
             }
 
-            if (openIDConnectConfigElem.getFirstChildWithName(getQNameWithIdentityNS(ConfigElements.ENCRYPTION_ALGORITHM)) != null) {
-                idTokenEncryptionAlgorithm =
-                        openIDConnectConfigElem.getFirstChildWithName(getQNameWithIdentityNS(ConfigElements.ENCRYPTION_ALGORITHM))
+            if (openIDConnectConfigElem.getFirstChildWithName(getQNameWithIdentityNS(ConfigElements.ID_TOKEN_ENCRYPTION_ALGORITHM)) != null) {
```
Comment (madurangasiriwardena) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/config/OAuthServerConfiguration.java]: Length seems to be exceeding 120 characters

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/config/OAuthServerConfiguration.java [hunk 12]
```diff
@@ -2155,6 +2188,50 @@ private void parseOpenIDConnectConfig(OMElement oauthConfigElem) {
...
+     */
+    private void parseSupportedIdTokenEncryptionAlgorithms(OMElement algorithms) {
+
+        if (algorithms == null) {
+            return;
+        }
+
+        Iterator iterator = algorithms.getChildrenWithLocalName(ConfigElements.SUPPORTED_ID_TOKEN_ENCRYPTION_ALGORITHM);
+        if (iterator != null) {
+            for (; iterator.hasNext(); ) {
```
Comment (madurangasiriwardena) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/config/OAuthServerConfiguration.java]: better to use while loop

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/config/OAuthServerConfiguration.java (hunk 11 above)
Comment (vihanga-liyanage) [reply components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/config/OAuthServerConfiguration.java]: Fixed with https://github.com/wso2-extensions/identity-inbound-auth-oauth/pull/960

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/config/OAuthServerConfiguration.java (hunk 12 above)
Comment (vihanga-liyanage) [reply components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/config/OAuthServerConfiguration.java]: Fixed with https://github.com/wso2-extensions/identity-inbound-auth-oauth/pull/960
